lib/watcher.ts

```diff
@@ -96,8 +96,8 @@ class Watcher extends EventEmitter {
 
     this._rebuildScheduled = true;
 
-    // Wait for current build, and ignore build failure
-    return Promise.resolve(this.currentBuild)
+    // If there is a current build, call cancel on it and ignore build failure
+    return Promise.resolve(this.builder.cancel())
```

Comment:
Are there cases where the cancellation needs to be communicated to plugins (i.e., if they have sidecar processes like a TS incremental compiler) so that they can take appropriate action? 

I'd like to understand context around important reasons to wait for `this.currentBuild` before proceeding with the next one.

Reply:
Hi @mike-north 👋 

>Are there cases where the cancellation needs to be communicated to plugins (i.e., if they have sidecar processes like a TS incremental compiler) so that they can take appropriate action?

Currently plugin creators can hook into the `cancel` hook to ensure that any existing work that needs to be cleaned up, can be cleaned up. 

Looking at the typescript compiler it looks like it doesn't have a cancel plugin. I can also see that the plugin is not returning a promise to build https://github.com/tildeio/broccoli-typescript-compiler/blob/master/src/plugin.ts#L51. 

Looking at the docs https://github.com/broccolijs/broccoli/blob/dfd95c07b7c1284f7a86945a1665c596345995a8/docs/node-api.md it says as a part of this that the build function must return a promise.

```
it must return a promise that is resolved on completion of the asynchronous work, or rejected if there is an error. Return values other than promises are ignored.
```

This code would not affect the typescript funnel as it is not cancellable from what I can infer.

> I'd like to understand context around important reasons to wait for this.currentBuild before proceeding with the next one.

This is primarily an issue for the project I am working on. Where you edit the code in multiple places and you must way (n) times for the build to finish. If a basic build takes 15 seconds and you edit 5 different places, you must wait 75 seconds. 

My main reason for adding this is why do we want for a build to succeed as that funnel is already waiting to build again. If we were to cancel the funnel, we would save time not having wait for stale builds to finish.

Reply:
@mike-north we don't current cancel running plugins, we simply let the current plugin complete but don't proceed to the next. This is already a reality plugins must be aware of as recoverable build errors will have the same behavior.